File: backend/src/schemas/task.py
from pydantic import BaseModel
from typing import Optional
from datetime import datetime
from enum import Enum

# Priority Interface
class PriorityLevel(str, Enum):
    low = "low"
    medium = "medium"
    high = "high"
    urgent = "urgent"


# Task Interface
class TaskCreate(BaseModel):
    """
    Schema for creating a new task.
    """
    title: str
    description: Optional[str] = None
    completed: bool = False
    priority: Optional[PriorityLevel] = None


# Task Update Interface
class TaskUpdate(BaseModel):
    """
    Schema for updating an existing task.
    """
    title: Optional[str] = None
    description: Optional[str] = None
    completed: Optional[bool] = None
    priority: Optional[PriorityLevel] = None


# Task Response Interface
class TaskResponse(BaseModel):
    """
    Schema for task response from API.
    """
    id: int
    title: str
    description: Optional[str]
    completed: bool
    priority: PriorityLevel
    user_id: int
    created_at: datetime
    updated_at: datetime

File: frontend/src/app/types/task.ts
// frontend/src/app/types/task.ts
export interface Task {
    id: string;
    user_id: string; // Assuming UUID from backend is string in frontend
    title: string;
    description?: string;
    status: 'pending' | 'completed'; // Matches backend enum
    priority: 'low' | 'medium' | 'high'; // Matches backend enum
    categories?: string[];
    is_recurring?: boolean;
    recurrence_pattern?: 'daily' | 'weekly' | 'monthly'; // Matches backend enum
    created_at: string; // Assuming datetime from backend is string in frontend
    due_date?: string; // Assuming datetime from backend is string in frontend
    updated_at: string; // Assuming datetime from backend is string in frontend
}

// Optional: Define DTOs for creation and update if they differ significantly
export interface TaskCreate {
    title: string;
    description?: string;
    due_date?: string;
    status?: 'pending' | 'completed';
    priority?: 'low' | 'medium' | 'high';
    categories?: string[];
    is_recurring?: boolean;
    recurrence_pattern?: 'daily' | 'weekly' | 'monthly';
}

export interface TaskUpdate {
    title?: string;
    description?: string;
    due_date?: string;
    status?: 'pending' | 'completed';
    priority?: 'low' | 'medium' | 'high';
    categories?: string[];
    is_recurring?: boolean;
    recurrence_pattern?: 'daily' | 'weekly' | 'monthly';
}
